peer_heap.js

```diff
@@ -104,13 +97,20 @@ PeerHeap.prototype.choose = function choose(threshold, filter) {
 
         var el = self.array[i];
 
-        if (self.rangehis[i] <= self.maxRangeStart) {
+        if (!el || (
+            self.rangehis[i] <= self.maxRangeStart &&
+            (!filter && !notEnoughPeers)
+        )) {
             // This range ends before the range with the largest start begins,
             // so it can't possibly be chosen over any of the ranges we've
             // seen. All ranges below this one have a smaller end, so this
             // range and any below it can't be chosen.
             continue;
-        } else if (!filter || filter(el.peer)) {
+        }
+
+        if (!filter || filter(el.peer)) {
+            isSecondary = notEnoughPeers && el.peer.isConnected('out');
```

Comment:
I did a lot of work to make sure this loop was only accessing small, packed arrays of numbers. Even reaching into an object like this will violate cache locality, and I'd really rather not have a function call that can't be inlined. Did you benchmark / flamegraph this?